content/posts/rust-pyo3-support-in-pyodide.md

```diff
@@ -0,0 +1,338 @@
+
+Later on, we saw the load-time error 
+```
+bad export type for `_emscripten_get_now`: undefined
+```
+PyO3 v0.15 depends on the <code class="pkg">instant</code> crate via the <code
+class="pkg">parking_lot</code> crate. The <code class="pkg">instant</code> crate
+tries to explicitly support Emscripten but it mispells `emscripten_get_now` as
+`_emscripten_get_now` which leads to linker errors. We have to patch <code
+class="pkg">instant</code> and use a Cargo dependency override. The <code
+class="pkg">instant</code> crate does not seem to be maintained anymore so [my
+find-and-replace patch](https://github.com/sebcrozet/instant/pull/47) has not
+been accepted. Luckily <code class="pkg">parking_lot</code> v0.12 does not use
+<code class="pkg">instant</code> anymore. Unfortunately, PyO3 v0.15 has a
+version pin on <code class="pkg">parking_lot</code> v0.11 and PyO3 v0.16 dropped
+support for Python 3.6 which Cryptography still supports. So Cryptography pins
+PyO3 to v0.15 and depends on <code class="pkg">instant</code> and we need a
+patch.
+
+
+## `lib.rmeta` and `--whole-archive`
+
+When attempting to build a dynamic Rust library with Emscripten, we see the
+linker error [`error: unknown file type:
+lib.rmeta`](https://github.com/rust-lang/rust/issues/80775).
+
+This error occurs because the `--whole-archive` option is present when linking.
+According to the GNU linker manual:
+```quote
+For each archive mentioned on the command line after the --whole-archive option,
+include every object file in the archive in the link, rather than searching the
+archive for the required object files.  This is normally used to turn an archive
+file into a shared library
+```
+In particular, `--whole-archive` requires every file in the library to be an
+object file. [Rust libraries contain an extra metadata file called
+`lib.rmeta`](https://rustc-dev-guide.rust-lang.org/backend/libs-and-metadata.html#rlib)
+hence the error message.
+
+Emscripten has two ways of generating a dynamic library: 
+
+* `-sSIDE_MODULE=1` exports all symbols by wrapping the linker arguments with
+  `--whole-archive`
+* `-sSIDE_MODULE=2` exports an explicit list of symbols 
+
+Building with `-sSIDE_MODULE=2` is much better for code size because symbols
+that are not exported can be inlined or eliminated as dead code during link-time
+optimization. But to use `-sSIDE_MODULE=2` we need to calculate the symbols to
+export somehow.
+
+Using `-sSIDE_MODULE=2` solves the `lib.rmeta` error because `--whole-archive`
+is not passed to the linker. Conveniently, Rust is good at calculating which
+symbols should be exported: public symbols with the `#[no_mangle]` attribute are
+exported, other symbols are not. In the case of a PyO3 module, the only exported
+symbol is the `PyInit__my_module` function that Python invokes when loading a
+native module. 
+
+It would be great to link our other extension modules with `-sSIDE_MODULE=2` but
```

Comment:
I'm a little unsure what this is aspiring to (what "our other extension modules" are).